### DISCUSSION TITLE
How to auth user in react application?

### DISCUSSION THREAD

**Discussion by FreePhoenix888** [deep-foundation/Discussions#10] [Q&A] [answered]
I pass gqlPath and token to Provider:
https://github.com/deep-foundation/sdk/blob/b4cfaeb2860b14bad90ea3a9cbe895f21d15bba2/pages/index.tsx#L216

Provider uses it as ApolloClientTokenizedProvider options
https://github.com/deep-foundation/sdk/blob/b4cfaeb2860b14bad90ea3a9cbe895f21d15bba2/imports/provider.tsx#L38-L43

Console logs:
https://github.com/deep-foundation/sdk/blob/b4cfaeb2860b14bad90ea3a9cbe895f21d15bba2/pages/index.tsx#L210-L212
![image](https://user-images.githubusercontent.com/66206278/236389444-690d0177-afb4-4162-96fa-80863d9b1ded.png)

Problem:
deep.token is not changed:
![image](https://user-images.githubusercontent.com/66206278/236389666-90b8de15-53a8-4a6e-ae38-0e03886067eb.png)

**Comment by FreePhoenix888** [comment]
There were no problems with apollo client path but token has not been changing

**I have fixed this problem by using `useTokenController` instead of managing token by using apollo client props**

Why do not we have hook to manage apollo client path? @Konard

**Reply by konard** [reply]
I don't know, ask @ivansglazunov

**Reply by FreePhoenix888** [reply]
> I don't know, ask @ivansglazunov

Do not we depend on @ivansglazunov too much, @Konard ?

**Reply by konard** [reply]
@FreePhoenix888 well you can try to create such hook if you want, and later discuss it with @ivansglazunov

**Reply by FreePhoenix888** [reply]
I have found the problem. Even if token is set for DeepClient, its linkId is still 0

**Reply by FreePhoenix888** [reply]
I can use deep.login instead of directly setting token

**Reply by FreePhoenix888** [reply, marked as answer]
Use deep.login and pass token there and pass gql path to
```ts
						<ApolloClientTokenizedProvider
							options={{
								client: 'deeplinks-app',
								path: gqlPath,
								ssl: isSsl,
								ws: !!process?.browser,
							}}
						>
```
Working code:
```ts
function LoginOrPage({ setGqlPath }) {
  const deep = useDeep();
  const [isAuthorized, setIsAuthorized] = useState(false);

  useEffect(() => {
    self["deep"] = deep
    if (deep.linkId !== 0) {
      setIsAuthorized(true);
    } else {
      setIsAuthorized(false);
    }
  }, [deep]);

  return isAuthorized ? (
    <Page />
  ) : (
    <LoginCard
      onSubmit={(arg) => {
        console.log({arg})
        setGqlPath(arg.gqlPath);
        deep.login({
          token: arg.token
        })
      }}
    />
  );
}

export default function Index() {
  const [gqlPath, setGqlPath] = useState<string | undefined>(undefined)

  useEffect(() => {
    console.log({gqlPath})
  }, [gqlPath])

  return (
    <>
      <Provider gqlPath={gqlPath} isSsl={true}>
        <LoginOrPage setGqlPath={setGqlPath} />
      </Provider>
    </>
  );
}
```

**Comment by konard** [comment]
@FreePhoenix888 Well such questions is better to directly publish in StackOverflow

**Comment by konard** [comment]
@FreePhoenix888 can I kindly ask you to create a bug that tries to set token using generateApolloClient and passing it to DeepProvider directly, and getting not token on request to links?
